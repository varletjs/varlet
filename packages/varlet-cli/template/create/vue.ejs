<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
