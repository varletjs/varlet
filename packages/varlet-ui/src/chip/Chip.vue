<template>
  <transition name="var-fade">
    <span :class="classes(n(), 'var--box', ...contentClass)" :style="chipStyles" v-bind="$attrs">
      <slot name="left" />

      <span :class="n(`text-${size}`)">
        <slot />
      </span>

      <slot name="right" />

      <span v-if="closable" :class="n('--close')" @click="onClose">
        <var-icon :name="`${iconName ? iconName : 'close-circle'}`" />
      </span>
    </span>
  </transition>
</template>

<script lang="ts">
import VarIcon from '../icon'
import { defineComponent, computed } from 'vue'
import { props } from './props'
import type { ComputedRef } from 'vue'
import { createNamespace } from '../utils/components'

const { n, classes } = createNamespace('chip')

export default defineComponent({
  name: 'VarChip',
  components: {
    VarIcon,
  },
  inheritAttrs: false,
  props,
  setup(props) {
    const chipStyles = computed(() => {
      const { plain, textColor, color } = props

      if (plain) {
        return {
          color: textColor || color,
          borderColor: color,
        }
      }

      return {
        color: textColor,
        background: color,
      }
    })

    const contentClass: ComputedRef<Array<string | false | undefined>> = computed(() => {
      const { size, block, type, plain, round } = props

      const blockClass = block ? 'var--flex' : 'var--inline-flex'
      const plainTypeClass = plain ? `${n('plain')} ${n(`plain-${type}`)}` : n(`--${type}`)
      const roundClass = round && n('--round')

      return [n(`--${size}`), blockClass, plainTypeClass, roundClass]
    })

    return {
      n,
      classes,
      chipStyles,
      contentClass,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import './chip';
</style>
