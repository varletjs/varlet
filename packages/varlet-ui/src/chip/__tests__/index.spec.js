import Chip from '..'
import VarChip from '../Chip'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'

test('test chip plugin', () => {
  const app = createApp({}).use(Chip)
  expect(app.component(Chip.name)).toBeTruthy()
})

test('test chip close', () => {
  const onClose = vi.fn()
  const wrapper = mount(VarChip, {
    props: {
      onClose,
      closable: true,
    },
  })
  const closeEl = wrapper.find('.var-chip--close')

  expect(closeEl.exists()).toBe(true)
  closeEl.trigger('click')
  expect(onClose).toHaveBeenCalledTimes(1)
})

describe('test chip component props', () => {
  test('test chip type', () => {
    ;['default', 'primary', 'info', 'success', 'warning', 'danger'].forEach((type) => {
      const wrapper = mount(VarChip, {
        props: { type },
      })

      expect(wrapper.find('.var-chip--' + type).exists()).toBe(true)
      wrapper.unmount()
    })
  })

  test('test chip size', () => {
    ;['normal', 'mini', 'small', 'large'].forEach((size) => {
      const wrapper = mount(VarChip, {
        props: { size },
      })

      expect(wrapper.find('.var-chip--' + size).exists()).toBe(true)
      wrapper.unmount()
    })
  })

  test('test chip plain', async () => {
    const wrapper = mount(VarChip, {
      props: {
        plain: true,
      },
    })

    expect(wrapper.find('.var-chip__plain').exists()).toBe(true)
    await wrapper.setProps({ plain: false })
    expect(wrapper.find('.var-chip__plain').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test chip round', async () => {
    const wrapper = mount(VarChip, {
      props: {
        round: true,
      },
    })

    expect(wrapper.find('.var-chip--round').exists()).toBe(true)
    await wrapper.setProps({ round: false })
    expect(wrapper.find('.var-chip--round').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test chip block', async () => {
    const wrapper = mount(VarChip, {
      props: {
        block: true,
      },
    })

    expect(wrapper.find('.var--flex').exists()).toBe(true)
    await wrapper.setProps({ block: false })
    expect(wrapper.find('.var--inline-flex').exists()).toBe(true)
    wrapper.unmount()
  })

  test('test chip closeable', async () => {
    const wrapper = mount(VarChip, {
      props: {
        closeable: true,
      },
    })

    expect(wrapper.find('.var-chip').attributes('closeable')).toContain(true)
    await wrapper.setProps({ closeable: false })
    expect(wrapper.find('.var-chip').attributes('closeable')).toContain(false)
    wrapper.unmount()
  })

  test('test chip close name', async () => {
    const wrapper = mount(VarChip, {
      props: {
        closeable: true,
        closeName: 'This is close name',
      },
    })

    expect(wrapper.find('.var-chip').attributes('closename')).toBe('This is close name')
    expect(wrapper.find('.var-chip').attributes('closeable')).toContain(true)
    await wrapper.setProps({ closeable: false })
    expect(wrapper.find('.var-chip').attributes('closeable')).toContain(false)
    wrapper.unmount()
  })

  test('test chip color', () => {
    const wrapper = mount(VarChip, {
      props: {
        color: 'red',
      },
    })

    expect(wrapper.find('.var-chip').attributes('style')).toContain('background: red;')
    wrapper.unmount()
  })

  test('test chip text color', () => {
    const wrapper = mount(VarChip, {
      props: {
        textColor: 'red',
      },
    })

    expect(wrapper.find('.var-chip').attributes('style')).toContain('color: red;')
    wrapper.unmount()
  })
})

describe('test chip component slots', () => {
  test('test chip default slots', () => {
    const wrapper = mount(VarChip, {
      slots: {
        default: () => 'This is default slots',
      },
    })

    expect(wrapper.find('.var-chip__text-normal').element.textContent).toBe('This is default slots')
    wrapper.unmount()
  })

  test('test chip left slots', () => {
    const wrapper = mount(VarChip, {
      slots: {
        left: () => 'This is left slots',
      },
    })

    expect(wrapper.find('.var-chip').html()).toContain('This is left slots<span class="var-chip__text-normal"></span>')
    wrapper.unmount()
  })

  test('test chip right slots', () => {
    const wrapper = mount(VarChip, {
      slots: {
        right: () => 'This is right slots',
      },
    })

    expect(wrapper.find('.var-chip').html()).toContain('<span class="var-chip__text-normal"></span>This is right slots')
    wrapper.unmount()
  })
})
