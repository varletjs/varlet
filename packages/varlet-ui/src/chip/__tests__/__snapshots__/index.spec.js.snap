// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test chip color & textColor 1`] = `
"<transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\" style=\\"color: rgb(0, 0, 0); background: rgb(0, 0, 0);\\"><span class=\\"var-chip__text-normal\\"></span>
  <!--v-if--></span>
</transition-stub>"
`;

exports[`test chip example 1`] = `
"<div class=\\"example\\">
  <div class=\\"app-type\\">纸片类型</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\"><span class=\\"var-chip__text-normal\\">默认纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--primary var-chip--round\\"><span class=\\"var-chip__text-normal\\">主要纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--success var-chip--round\\"><span class=\\"var-chip__text-normal\\">成功纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--danger var-chip--round\\"><span class=\\"var-chip__text-normal\\">危险纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--warning var-chip--round\\"><span class=\\"var-chip__text-normal\\">警告纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--info var-chip--round\\"><span class=\\"var-chip__text-normal\\">信息纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <div class=\\"app-type\\">简单纸片</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-primary var-chip--round\\"><span class=\\"var-chip__text-normal\\">简单纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <div class=\\"app-type\\">非圆角纸片</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--primary\\"><span class=\\"var-chip__text-normal\\">非圆角纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <div class=\\"app-type\\">纸片尺寸</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--primary var-chip--round\\"><span class=\\"var-chip__text-normal\\">常规纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--small var--inline-flex var-chip--success var-chip--round\\"><span class=\\"var-chip__text-small\\">小型纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--mini var--inline-flex var-chip--warning var-chip--round\\"><span class=\\"var-chip__text-mini\\">迷你纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--large var--inline-flex var-chip--danger var-chip--round\\"><span class=\\"var-chip__text-large\\">大型纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <div class=\\"app-type\\">块级纸片</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--flex var-chip--primary var-chip--round\\"><span class=\\"var-chip__text-normal\\">块级纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <div class=\\"app-type\\">可关闭纸片</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\"><span class=\\"var-chip__text-normal\\">可关闭纸片</span><span class=\\"var-chip--close\\"><i class=\\"var-icon var-icon--set var-icon-close-circle\\" style=\\"transition: transform 0ms;\\"></i></span></span></transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\"><span class=\\"var-chip__text-normal\\">自定义关闭图标</span><span class=\\"var-chip--close\\"><i class=\\"var-icon var-icon--set var-icon-delete\\" style=\\"transition: transform 0ms;\\"></i></span></span></transition-stub>
  <div class=\\"app-type\\">自定义颜色</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\" style=\\"color: rgb(255, 255, 255); background: rgb(0, 150, 136);\\"><span class=\\"var-chip__text-normal\\">纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default var-chip--round\\" style=\\"color: rgb(0, 150, 136); border-color: #009688;\\"><span class=\\"var-chip__text-normal\\">纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip__plain var-chip__plain-default var-chip--round\\" style=\\"color: rgb(230, 162, 60); border-color: #faecd8;\\"><span class=\\"var-chip__text-normal\\">纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\" style=\\"color: rgb(230, 162, 60); background: rgb(250, 236, 216);\\"><span class=\\"var-chip__text-normal\\">纸片</span>
    <!--v-if--></span>
  </transition-stub>
  <div class=\\"app-type\\">添加插槽</div>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\"><i class=\\"var-icon var-icon--set var-icon-star\\" style=\\"transition: transform 0ms;\\"></i><span class=\\"var-chip__text-normal\\">左侧插槽 </span>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\"><span class=\\"var-chip__text-normal\\">右侧插槽 </span><i class=\\"var-icon var-icon--set var-icon-fire\\" style=\\"transition: transform 0ms;\\"></i>
    <!--v-if--></span>
  </transition-stub>
  <transition-stub><span class=\\"var-chip var--box var-chip--normal var--inline-flex var-chip--default var-chip--round\\"><i class=\\"var-icon var-icon--set var-icon-account-circle\\" style=\\"transition: transform 0ms;\\"></i><span class=\\"var-chip__text-normal\\">左右两侧插槽 </span><i class=\\"var-icon var-icon--set var-icon-cake-variant\\" style=\\"transition: transform 0ms;\\"></i>
    <!--v-if--></span>
  </transition-stub>
</div>"
`;

exports[`test chip styles 1`] = `"<transition-stub><span class=\\"var-chip var--box var-chip--large var--flex var-chip__plain var-chip__plain-default var-chip--round\\"><span class=\\"var-chip__text-large\\">chip text</span><span class=\\"var-chip--close\\"><i class=\\"var-icon var-icon--set var-icon-fire\\" style=\\"transition: transform 0ms;\\"></i></span></span></transition-stub>"`;
